e2e/tenant_control_plane_additional_resources_ready_test.go

```diff
@@ -0,0 +1,339 @@
+		Expect(k8sClient.Create(context.Background(), controllerManagerConfigMap)).NotTo(HaveOccurred())
+		Expect(k8sClient.Create(context.Background(), schedulerConfigMap)).NotTo(HaveOccurred())
+		Expect(k8sClient.Create(context.Background(), tcp)).NotTo(HaveOccurred())
+	})
+
+	// Delete the TenantControlPlane resource after test is finished
+	JustAfterEach(func() {
+		Expect(k8sClient.Delete(context.Background(), tcp)).Should(Succeed())
+		Expect(k8sClient.Delete(context.Background(), apiServerConfigMap)).NotTo(HaveOccurred())
+		Expect(k8sClient.Delete(context.Background(), controllerManagerConfigMap)).NotTo(HaveOccurred())
+		Expect(k8sClient.Delete(context.Background(), schedulerConfigMap)).NotTo(HaveOccurred())
+	})
+
+	It("should have the additional resources", func() {
+		// Should be ready
+		StatusMustEqualTo(tcp, kamajiv1alpha1.VersionReady)
+		// Should have a TCP deployment
+		deploy := appsv1.Deployment{}
+		Expect(k8sClient.Get(context.Background(), types.NamespacedName{
+			Name:      tcpName,
+			Namespace: namespace,
+		}, &deploy)).NotTo(HaveOccurred())
+		// Should have the configured AdditionalInitContainers
+		found, _ := utilities.HasNamedContainer(deploy.Spec.Template.Spec.InitContainers, initContainerName)
+		Expect(found).To(BeTrue())
+		// Should have the configured AdditionalContainers"
+		found, _ = utilities.HasNamedContainer(deploy.Spec.Template.Spec.Containers, additionalContainerName)
+		Expect(found).To(BeTrue())
+		//
+		// Should have the configured AdditionalVolumes and AdditionalVolumeMounts
+		//
+		// apiserver
+		found, _ = utilities.HasNamedVolume(deploy.Spec.Template.Spec.Volumes, apiServerVolumeName)
+		Expect(found).To(BeTrue())
+
+		found, containerIndex := utilities.HasNamedContainer(deploy.Spec.Template.Spec.Containers, "kube-apiserver")
+		Expect(found).To(BeTrue())
+
+		found, _ = utilities.HasNamedVolumeMount(deploy.Spec.Template.Spec.Containers[containerIndex].VolumeMounts, apiServerVolumeName)
+		Expect(found).To(BeTrue())
+		// scheduler
+		found, _ = utilities.HasNamedVolume(deploy.Spec.Template.Spec.Volumes, schedulerVolumeName)
+		Expect(found).To(BeTrue())
+
+		found, _ = utilities.HasNamedContainer(deploy.Spec.Template.Spec.Containers, "kube-scheduler")
+		Expect(found).To(BeTrue())
+
+		found, _ = utilities.HasNamedVolumeMount(deploy.Spec.Template.Spec.Containers[containerIndex].VolumeMounts, schedulerVolumeName)
+		Expect(found).To(BeTrue())
+		// controller-manager
+		found, _ = utilities.HasNamedVolume(deploy.Spec.Template.Spec.Volumes, controllerManagerVolumeName)
+		Expect(found).To(BeTrue())
+
+		found, _ = utilities.HasNamedContainer(deploy.Spec.Template.Spec.Containers, "kube-controller-manager")
+		Expect(found).To(BeTrue())
+
+		found, _ = utilities.HasNamedVolumeMount(deploy.Spec.Template.Spec.Containers[3].VolumeMounts, controllerManagerVolumeName)
+		Expect(found).To(BeTrue())
+
+		By("removing the additional init containers", func() {
```

Comment:
I wonder if it would make sense to have something like
```
By("removing the additional init containers", func() {
Eventually(func() error {
	Expect(k8sClient.Get(context.Background(), types.NamespacedName{Name: tcp.Name, Namespace: tcp.Namespace}, tcp)).NotTo(HaveOccurred())

	tcp.Spec.ControlPlane.Deployment.AdditionalInitContainers = nil
	tcp.Spec.ControlPlane.Deployment.AdditionalContainers = nil
	tcp.Spec.ControlPlane.Deployment.AdditionalVolumeMounts = nil
	tcp.Spec.ControlPlane.Deployment.AdditionalVolumes = nil

	return k8sClient.Update(context.Background(), tcp)
}, 10*time.Second, time.Second).ShouldNot(HaveOccurred())
```
and then maybe have some helper functions like `StatusMustEqualTo` to make the test a little more easy to read.
What do you think?

Reply:
Due to resource versions update there's a lot of repeated code in the e2e suite, such as:

```go
Eventually(func() error {
  Expect(k8sClient.Get(context.Background(), types.NamespacedName{Name: tcp.Name, Namespace: tcp.Namespace}, tcp)).NotTo(HaveOccurred())

  // do you changes

  return k8sClient.Update(context.Background(), tcp)
}, 10*time.Second, time.Second).ShouldNot(HaveOccurred())
```

We could try to provide a syntactic sugar to hide this complexity but I would address this in a different PR.

Having a single `By` group should be enough, tho, since there's no correlation between additional (init) containers and volumes.

I'm gonna fix the last commit about tests.